src/type.js

```diff
@@ -459,6 +441,30 @@ Sk.builtin.type = function (name, bases, dict) {
             }
         }
 
+        // tp$getattr is a special case; we need to catch AttributeErrors and
+        // return undefined instead.
+        let getattributeFn = Sk.builtin.type.typeLookup(klass, Sk.builtin.str.$getattribute);
+        if (getattributeFn !== undefined && getattributeFn !== Sk.builtin.object.prototype.__getattribute__) {
+            klass.prototype.tp$getattr = function (pyName, canSuspend) {
+                let r = Sk.misceval.tryCatch(
+                    () => Sk.misceval.callsimOrSuspendArray(getattributeFn, [this, pyName]),
+                    function (e) {
+                        if (e instanceof Sk.builtin.AttributeError) {
+                            return undefined;
+                        } else {
+                            throw e;
+                        }
+                    }
+                );
+                return canSuspend ? r : Sk.misceval.retryOptionalSuspensionOrThrow(r);
+            };
+        } else if (!klass.prototype.tp$getattr) {
+            // This is only relevant in Python 2, where
+            // it's possible not to inherit from object
+            // (or perhaps when inheriting from builtins? Unclear)
+            klass.prototype.tp$getattr = Sk.builtin.object.prototype.GenericGetAttr;
+        }
+
```

Comment:
What happens if you monkey patch the class to add a ```__getatttribute__``` method here?  Personally, I don't care, because you deserve what you get (in my opinion!) if you do that, but does that create a correctness issue?

Reply:
You're right; it now breaks, like every other dunder-method (monkey-patching of `__len__`, `__getitem__`, or others isn't reflected in the JS slot functions either).

I can imagine a custom `tp$setattr` for classes that updates the prototype if you do a monkey-patch like that...although that won't work if you do it to a multiply-inherited ancestor (JS prototypical inheritance does not follow the Python MRO!). Subclasses could register a callback to be notified if it happens, I suppose, but that makes subclasses non-garbage-collectable, which I'm not too keen on.

Or we could set a global "someone monkey-patched something important" flag, which causes _all_ slot functions to deoptimise forevermore...although that doesn't help if the slot functions aren't even defined (which is the case for `sq$len` and friends).

Basically, it's a gnarly one to support, and our support for all the _other_ slot functions is already intolerant of monkey-patching, so I'm inclined to let it stay broken unless we can work out how to fix them all at once.